File: dotnet/src/Botas/BotApplication.cs
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using System.Collections;
using System.Text.Json;

namespace Botas;

/// <summary>
/// Exception thrown when an activity handler or invoke handler raises an unhandled error.
/// Wraps the original exception and includes the <see cref="Activity"/> that caused the failure.
/// </summary>
public class BotHandlerException(string message, Exception ex, CoreActivity activity) : Exception(message, ex)
{
    /// <summary>The activity that was being processed when the error occurred.</summary>
    public CoreActivity Activity { get; } = activity;
}

/// <summary>
/// Response returned by an invoke activity handler.
/// The <see cref="Status"/> is written as the HTTP status code;
/// <see cref="Body"/> is serialized as JSON.
/// </summary>
public class InvokeResponse
{
    /// <summary>HTTP status code to return to the channel (e.g. 200, 400, 501).</summary>
    public int Status { get; set; }
    /// <summary>Optional response body serialized as JSON. Omitted when <c>null</c>.</summary>
    public object? Body { get; set; }
}

/// <summary>
/// Delegate that advances the middleware pipeline to the next middleware or the final handler.
/// </summary>
/// <param name="cancellationToken">Token to cancel the pipeline execution.</param>
public delegate Task NextDelegate(CancellationToken cancellationToken);

/// <summary>
/// Interface for middleware components that participate in the bot turn pipeline.
/// Middleware is invoked in registration order before the activity handler.
/// </summary>
public interface ITurnMiddleWare
{
    /// <summary>
    /// Processes the turn and optionally calls <paramref name="next"/> to continue the pipeline.
    /// Not calling <paramref name="next"/> short-circuits the remaining middleware and handler.
    /// </summary>
    /// <param name="context">The turn context for the current activity.</param>
    /// <param name="next">Delegate to invoke the next middleware or handler in the pipeline.</param>
    /// <param name="cancellationToken">Token to cancel processing.</param>
    Task OnTurnAsync(TurnContext context, NextDelegate next, CancellationToken cancellationToken = default);
}

/// <summary>
/// Core bot application that processes incoming activities through a middleware pipeline
/// and dispatches them to registered handlers. This is the central entry point for the Botas SDK.
/// </summary>
/// <remarks>
/// Activities arrive via <see cref="ProcessAsync"/>, pass through JWT validation (handled externally
/// by ASP.NET Core authentication), then flow through registered middleware in order, and finally
/// reach the appropriate handler based on <c>activity.type</c>.
/// </remarks>
public class BotApplication
{
    /// <summary>The Botas SDK version.</summary>
    public static string Version => ThisAssembly.AssemblyInformationalVersion;

    private readonly ILogger<BotApplication> _logger;
    private readonly IConfiguration _configuration;
    private ConversationClient? _conversationClient;
    private readonly string _serviceKey;
    private readonly TurnMiddleware _turnMiddleware;
    private readonly Dictionary<string, Func<TurnContext, CancellationToken, Task>> _handlers = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, Func<TurnContext, CancellationToken, Task<InvokeResponse>>> _invokeHandlers = new(StringComparer.OrdinalIgnoreCase);

    /// <summary>
    /// Initializes a new <see cref="BotApplication"/> with no-op logging and empty configuration.
    /// Intended for unit testing scenarios.
    /// </summary>
    public BotApplication()
    {
        _logger = NullLogger<BotApplication>.Instance;
        _configuration = new ConfigurationBuilder().Build();
        _serviceKey = "AzureAd";
        _turnMiddleware = new TurnMiddleware();
    }

    /// <summary>
    /// Initializes a new <see cref="BotApplication"/> with the specified configuration and logger.
    /// </summary>
    /// <param name="config">Application configuration (used to resolve <c>ClientId</c> and other settings).</param>
    /// <param name="logger">Logger instance for diagnostic output.</param>
    /// <param name="serviceKey">The configuration section name for Azure AD settings (default: <c>"AzureAd"</c>).</param>
    public BotApplication(IConfiguration config, ILogger<BotApplication> logger, string serviceKey = "AzureAd")
    {
        _logger = logger;
        _configuration = config;
        _serviceKey = serviceKey;
        _turnMiddleware = new TurnMiddleware();
        logger.LogInformation("Started bot listener on {Port} for AppID:{AppId}", config["ASPNETCORE_URLS"], config[$"{_serviceKey}:ClientId"]);
    }

    internal TurnMiddleware MiddleWare => _turnMiddleware;

    /// <summary>
    /// Optional catch-all activity handler. When set, bypasses all per-type handlers registered via <see cref="On"/>
    /// and receives every activity regardless of type.
    /// </summary>
    public Func<TurnContext, CancellationToken, Task>? OnActivity { get; set; }

    /// <summary>
    /// The Azure AD application (client) ID for this bot, read from configuration. Returns <c>null</c> if not configured.
    /// </summary>
    public string? AppId => _configuration[$"{_serviceKey}:ClientId"];

    /// <summary>
    /// Register a handler for a specific activity type.
    /// Only one handler per type is supported; registering the same type replaces the previous handler.
    /// </summary>
    public BotApplication On(string type, Func<TurnContext, CancellationToken, Task> handler)
    {
        _handlers[type] = handler;
        return this;
    }

    /// <summary>
    /// Register a handler for an invoke activity by its <c>activity.Name</c> sub-type.
    /// The handler must return an <see cref="InvokeResponse"/> — the status and body are
    /// written directly to the HTTP response for invoke activities.
    /// Only one handler per name is supported; registering the same name replaces the previous handler.
    /// </summary>
    public BotApplication OnInvoke(string name, Func<TurnContext, CancellationToken, Task<InvokeResponse>> handler)
    {
        _invokeHandlers[name] = handler;
        return this;
    }

    /// <summary>
    /// Processes an incoming HTTP request containing a Bot Service activity.
    /// Deserializes the activity, runs it through the middleware pipeline and handler dispatch,
    /// and writes the HTTP response (JSON <c>{}</c> on success, or an <see cref="InvokeResponse"/> for invoke activities).
    /// </summary>
    /// <param name="httpContext">The ASP.NET Core HTTP context containing the incoming request.</param>
    /// <param name="cancellationToken">Token to cancel processing.</param>
    /// <returns>The deserialized <see cref="CoreActivity"/> that was processed.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the activity is missing required fields or the <see cref="ConversationClient"/> is not registered.</exception>
    /// <exception cref="BotHandlerException">Thrown when a handler raises an unhandled exception.</exception>
    public async Task<CoreActivity> ProcessAsync(HttpContext httpContext, CancellationToken cancellationToken = default)
    {
        _conversationClient = httpContext.RequestServices.GetKeyedService<ConversationClient>(_serviceKey) ?? throw new InvalidOperationException("ConversationClient not registered");

        CoreActivity activity = await CoreActivity.FromJsonStreamAsync(httpContext.Request.Body, cancellationToken) ?? throw new InvalidOperationException("Invalid Activity");

        if (string.IsNullOrEmpty(activity.Type))
        {
            throw new InvalidOperationException("Activity Type is required");
        }
        if (activity.Conversation?.Id is null)
        {
            throw new InvalidOperationException("Activity Conversation.Id is required");
        }
        if (string.IsNullOrEmpty(activity.ServiceUrl))
        {
            throw new InvalidOperationException("Activity ServiceUrl is required");
        }

        if (_logger.IsEnabled(LogLevel.Trace))
        {
            _logger.LogTrace("Received activity type: {Type}", activity.Type);
        }

        using (_logger.BeginScope("Processing activity {Type}", activity.Type))
        {
            var context = new TurnContext(this, activity);
            InvokeResponse? invokeResponse = null;
            try
            {
                var callback = OnActivity is not null
                    ? (Func<TurnContext, CancellationToken, Task>)((ctx, ct) => OnActivity(ctx, ct))
                    : DispatchToHandler;

                if (string.Equals(activity.Type, "invoke", StringComparison.OrdinalIgnoreCase))
                {
                    Task<InvokeResponse> invokeCallback(TurnContext ctx, CancellationToken ct) => DispatchInvokeHandler(ctx, ct);
                    await _turnMiddleware.RunPipeline(context, async (ctx, ct) =>
                    {
                        invokeResponse = await invokeCallback(ctx, ct).ConfigureAwait(false);
                    }, 0, cancellationToken).ConfigureAwait(false);
                }
                else
                {
                    await _turnMiddleware.RunPipeline(context, callback, 0, cancellationToken).ConfigureAwait(false);
                }
            }
            catch (Exception ex)
            {
                if (ex is OperationCanceledException)
                {
                    throw;
                }
                if (ex is BotHandlerException)
                {
                    throw;
                }
                _logger.LogError(ex, "Error processing activity {Type}", activity.Type);
                throw new BotHandlerException("Error processing activity", ex, activity);
            }
            finally
            {
                _logger.LogInformation("Finished processing activity {Type}", activity.Type);
            }

            // Write the HTTP response — ProcessAsync owns the full response lifecycle
            if (invokeResponse is not null)
            {
                httpContext.Response.StatusCode = invokeResponse.Status;
                if (invokeResponse.Body is not null)
                {
                    httpContext.Response.ContentType = "application/json";
                    await httpContext.Response.WriteAsync(
                        JsonSerializer.Serialize(invokeResponse.Body, CoreActivity.DefaultJsonOptions),
                        cancellationToken).ConfigureAwait(false);
                }
            }
            else
            {
                httpContext.Response.StatusCode = StatusCodes.Status200OK;
                httpContext.Response.ContentType = "application/json";
                await httpContext.Response.WriteAsync("{}", cancellationToken).ConfigureAwait(false);
            }

            return activity;
        }
    }

    /// <summary>
    /// Register middleware to run before handlers on every turn.
    /// Middleware executes in registration order.
    /// IMPORTANT: Call this method only during application startup, not per-request.
    /// </summary>
    public ITurnMiddleWare Use(ITurnMiddleWare middleware)
    {
        _turnMiddleware.Use(middleware);
        return _turnMiddleware;
    }

    /// <summary>
    /// Sends an outbound activity to the Bot Service channel via the <see cref="ConversationClient"/>.
    /// </summary>
    /// <param name="activity">The activity to send. Must have routing fields (<c>ServiceUrl</c>, <c>Conversation</c>) populated.</param>
    /// <param name="cancellationToken">Token to cancel the send operation.</param>
    /// <returns>The raw JSON response from the Bot Service service.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the <see cref="ConversationClient"/> has not been initialized (no request has been processed yet).</exception>
    public async Task<string> SendActivityAsync(CoreActivity activity, CancellationToken cancellationToken = default)
    {
        if (_conversationClient is null)
        {
            throw new InvalidOperationException("ConversationClient not initialized");
        }
        return await _conversationClient.SendActivityAsync(activity, cancellationToken);
    }

    private Task DispatchToHandler(TurnContext context, CancellationToken cancellationToken)
    {
        if (_handlers.TryGetValue(context.Activity.Type, out var handler))
        {
            return handler(context, cancellationToken);
        }
        return Task.CompletedTask;
    }

    internal async Task<InvokeResponse> DispatchInvokeHandler(TurnContext context, CancellationToken cancellationToken)
    {
        if (_invokeHandlers.Count == 0)
        {
            return new InvokeResponse { Status = 200 };
        }

        var name = context.Activity.Name;
        if (name is not null && _invokeHandlers.TryGetValue(name, out var handler))
        {
            try
            {
                return await handler(context, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                throw new BotHandlerException($"Invoke handler for \"{name}\" threw an error", ex, context.Activity);
            }
        }
        return new InvokeResponse { Status = 501 };
    }
}

internal class TurnMiddleware : ITurnMiddleWare, IEnumerable<ITurnMiddleWare>
{

    private readonly IList<ITurnMiddleWare> _middlewares = [];
    internal TurnMiddleware Use(ITurnMiddleWare middleware)
    {
        _middlewares.Add(middleware);
        return this;
    }


    public async Task OnTurnAsync(TurnContext context, NextDelegate next, CancellationToken cancellationToken = default)
    {
        await RunPipeline(context, null!, 0, cancellationToken).ConfigureAwait(false);
        await next(cancellationToken).ConfigureAwait(false);
    }

    public Task RunPipeline(TurnContext context, Func<TurnContext, CancellationToken, Task>? callback, int nextMiddlewareIndex, CancellationToken cancellationToken)
    {
        if (nextMiddlewareIndex == _middlewares.Count)
        {
            if (callback is not null)
            {
                return callback!(context, cancellationToken) ?? Task.CompletedTask;
            }
            else
            {
                return Task.CompletedTask;
            }
        }
        ITurnMiddleWare nextMiddleware = _middlewares[nextMiddlewareIndex];
        return nextMiddleware.OnTurnAsync(
            context,
            (ct) => RunPipeline(context, callback, nextMiddlewareIndex + 1, ct),
            cancellationToken);

    }

    public IEnumerator<ITurnMiddleWare> GetEnumerator()
    {
        return _middlewares.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}

File: node/packages/botas-core/src/bot-application.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { IncomingMessage, ServerResponse } from 'node:http'
import { Buffer } from 'node:buffer'
import type { CoreActivity, ResourceResponse } from './core-activity.js'
import type { TurnMiddleware } from './turn-middleware.js'
import type { TurnContext } from './turn-context.js'
import { createTurnContext } from './turn-context.js'
import { ConversationClient } from './conversation-client.js'

import { TokenManager } from './token-manager.js'
import type { BotApplicationOptions } from './bot-application-options.js'
import { getLogger } from './logger.js'
import { validateServiceUrl } from './bot-auth-middleware.js'
import { withActiveSpan } from './tracer-provider.js'
import { getMetrics } from './meter-provider.js'
import { VERSION } from './version.js'

/** A function that handles a specific activity type. */
export type CoreActivityHandler = (context: TurnContext) => Promise<void>

/**
 * The response returned by an invoke activity handler.
 * The `status` is written as the HTTP status code; `body` is serialized as JSON.
 *
 * @public
 */
export interface InvokeResponse {
  /** HTTP status code to return to the channel (e.g. `200`, `400`, `501`). */
  status: number
  /** Optional response body serialized as JSON. */
  body?: unknown
}

/** A function that handles an invoke activity and returns an InvokeResponse. */
export type InvokeActivityHandler = (context: TurnContext) => Promise<InvokeResponse>

/**
 * Wraps an exception thrown by an activity handler with the originating activity.
 *
 * Catch this in error-handling middleware or at the HTTP layer to inspect both the
 * original error and the activity that triggered it.
 */
export class BotHandlerException extends Error {
  /**
   * @param message - Human-readable error description.
   * @param cause - The original error thrown by the handler.
   * @param activity - The activity being processed when the error occurred.
   */
  constructor (
    message: string,
    public override readonly cause: unknown,
    public readonly activity: CoreActivity
  ) {
    super(message)
    this.name = 'BotHandlerException'
  }
}

/**
 * Core bot application that processes incoming Bot Service activities.
 *
 * Web-server-agnostic — use {@link processAsync} with Express (or any
 * Node.js `http.Server`) and {@link processBody} with frameworks that manage
 * their own HTTP response lifecycle (e.g. Hono).
 *
 * @example
 * ```ts
 * const bot = new BotApplication({ clientId: '...', clientSecret: '...' });
 *
 * bot.on('message', async ({ activity, send }) => {
 *   await send(`You said: ${activity.text}`);
 * });
 * ```
 */
export class BotApplication {
  /** The Botas SDK version. */
  static readonly version: string = VERSION

  /** Resolved options for this application instance. */
  readonly options: BotApplicationOptions

  /** Client for sending, updating, and deleting activities via the Bot Service API. */
  readonly conversationClient: ConversationClient

  /**
   * Optional CatchAll handler. When set, completely replaces per-type handler
   * dispatch — all activities are delivered directly to this handler.
   */
  onActivity?: CoreActivityHandler

  private readonly middlewares: TurnMiddleware[] = []
  private readonly handlers = new Map<string, CoreActivityHandler>()
  private readonly invokeHandlers = new Map<string, InvokeActivityHandler>()
  private readonly tokenManager: TokenManager

  /**
   * Create a new BotApplication.
   *
   * Options fall back to environment variables (`CLIENT_ID`, `CLIENT_SECRET`,
   * `TENANT_ID`) when not provided explicitly.
   *
   * @param options - Bot credentials and configuration. Defaults to env vars.
   */
  constructor (options: BotApplicationOptions = {}) {
    const resolvedOptions = resolveBotApplicationOptions(options)
    this.options = resolvedOptions
    this.tokenManager = new TokenManager(resolvedOptions)
    const tokenProvider = () =>
      this.tokenManager.getBotToken().then((t) => {
        if (!t) throw new Error('No credentials configured — set CLIENT_ID and CLIENT_SECRET (or use managed identity)')
        return t
      })
    this.conversationClient = new ConversationClient(tokenProvider)
  }

  /**
   * Register a handler for a given activity type.
   *
   * Only one handler per type is supported; registering the same type twice
   * replaces the previous handler.
   *
   * @param type - CoreActivity type string (e.g. `"message"`, `"conversationUpdate"`).
   * @param handler - Async function called with the turn context.
   * @returns `this` for method chaining.
   */
  on (type: string, handler: CoreActivityHandler): this {
    this.handlers.set(type.toLowerCase(), handler)
    return this
  }

  /**
   * Add a middleware to the turn processing pipeline.
   *
   * Middleware runs in registration order before the activity handler.
   *
   * @param middleware - Middleware function to add. See {@link TurnMiddleware}.
   * @returns `this` for method chaining.
   */
  use (middleware: TurnMiddleware): this {
    this.middlewares.push(middleware)
    return this
  }

  /**
   * Register a handler for an invoke activity by its `activity.name` sub-type.
   *
   * The handler must return an {@link InvokeResponse} — the status and body are
   * written directly to the HTTP response for invoke activities.
   *
   * Only one handler per name is supported; registering the same name twice
   * replaces the previous handler.
   *
   * @param name - The invoke name (e.g. `"adaptiveCard/action"`, `"task/fetch"`).
   * @param handler - Async function that returns an {@link InvokeResponse}.
   * @returns `this` for method chaining.
   */
  onInvoke (name: string, handler: InvokeActivityHandler): this {
    this.invokeHandlers.set(name.toLowerCase(), handler)
    return this
  }

  /**
   * Process an incoming HTTP request (Express / Node.js `http.Server`).
   *
   * Reads the request body, validates the activity, runs the middleware
   * pipeline, and writes a `200 {}` response. For `invoke` activities,
   * writes the `InvokeResponse` returned by the registered handler instead.
   * Writes `413` for oversized bodies and `500` on unhandled errors.
   *
   * @param req - Node.js incoming HTTP request.
   * @param res - Node.js server response to write the result to.
   * @throws Never — errors are caught and written to `res` as HTTP status codes.
   */
  async processAsync (req: IncomingMessage, res: ServerResponse): Promise<void> {
    getLogger().debug('Start processing HTTP request for activity')
    try {
      const body = await readBody(req)
      const invokeResponse = await this.processBody(body)
      if (invokeResponse) {
        res.writeHead(invokeResponse.status, { 'Content-Type': 'application/json' })
        res.end(invokeResponse.body !== undefined ? JSON.stringify(invokeResponse.body) : '{}')
      } else {
        res.writeHead(200, { 'Content-Type': 'application/json' })
        res.end('{}')
      }
    } catch (err) {
      if (res.writableEnded || res.destroyed) return
      if (err instanceof RequestBodyTooLargeError) {
        res.writeHead(413)
        res.end(err.message)
      } else if (err instanceof ActivityValidationError) {
        res.writeHead(400, { 'Content-Type': 'application/json' })
        res.end(JSON.stringify({ error: err.message }))
      } else {
        getLogger().error('processAsync returning 500: %s', err instanceof Error ? err.message : String(err))
        if (!res.headersSent) {
          res.writeHead(500)
        }
        res.end('Internal server error')
      }
    }
  }

  /**
   * Process a raw JSON request body without touching the HTTP layer.
   *
   * Use this with frameworks that manage their own response lifecycle (e.g.
   * Hono), where writing to `ServerResponse` directly would cause issues.
   *
   * For `invoke` activities, the returned {@link InvokeResponse} must be
   * written to the HTTP response by the caller.
   *
   * @param body - Raw JSON string from the HTTP request body.
   * @returns The {@link InvokeResponse} for `invoke` activities, or
   *   `undefined` for all other activity types.
   * @throws {Error} If the body is not valid JSON or fails activity validation.
   * @throws {BotAuthError} If the activity's `serviceUrl` is not on the allowlist.
   * @throws {BotHandlerException} If an activity handler throws during processing.
   */
  async processBody (body: string): Promise<InvokeResponse | undefined> {
    const activity = safeJsonParse(body) as CoreActivity
    assertCoreActivity(activity)
    validateServiceUrl(activity.serviceUrl)
    getLogger().info('CoreActivity received: type=%s serviceUrl=%s', activity.type, activity.serviceUrl)
    getLogger().trace('Received activity: %s', body)

    const metrics = getMetrics()
    metrics?.activitiesReceived.add(1, { 'activity.type': activity.type })
    const startTime = Date.now()

    return withActiveSpan('botas.turn', async (turnSpan) => {
      turnSpan?.setAttribute('activity.type', activity.type)
      turnSpan?.setAttribute('activity.id', activity.id ?? '')
      turnSpan?.setAttribute('conversation.id', activity.conversation?.id ?? '')
      turnSpan?.setAttribute('channel.id', activity.channelId ?? '')
      turnSpan?.setAttribute('bot.id', this.options.clientId ?? '')

      try {
        const invokeResponse = await this.runPipelineAsync(activity)
        getLogger().info('Finished processing activity: type=%s', activity.type)
        return invokeResponse
      } catch (err) {
        getLogger().error('Error processing activity: type=%s', activity.type, err)
        throw err
      } finally {
        metrics?.turnDuration.record(Date.now() - startTime, { 'activity.type': activity.type })
      }
    })
  }

  /**
   * Proactively send an activity to a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param activity - CoreActivity payload to send.
   * @returns The {@link ResourceResponse} containing the new activity ID, or `undefined`.
   */
  async sendActivityAsync (
    serviceUrl: string,
    conversationId: string,
    activity: Partial<CoreActivity>
  ): Promise<ResourceResponse | undefined> {
    return this.conversationClient.sendCoreActivityAsync(serviceUrl, conversationId, activity)
  }

  /** @internal Dispatch the activity to the CatchAll or per-type handler. */
  protected async handleCoreActivityAsync (context: TurnContext): Promise<InvokeResponse | undefined> {
    if (context.activity.type === 'invoke') {
      return this.dispatchInvokeAsync(context)
    }

    const isCatchAll = !!this.onActivity
    const handler = this.onActivity ?? this.handlers.get(context.activity.type.toLowerCase())
    if (handler) {
      return withActiveSpan('botas.handler', async (handlerSpan) => {
        handlerSpan?.setAttribute('handler.type', context.activity.type)
        handlerSpan?.setAttribute('handler.dispatch', isCatchAll ? 'catchall' : 'type')
        try {
          await handler(context)
        } catch (err) {
          getMetrics()?.handlerErrors.add(1, { 'activity.type': context.activity.type })
          throw new BotHandlerException(
            `Handler for "${context.activity.type}" threw an error`,
            err,
            context.activity
          )
        }
        return undefined
      })
    }
    return undefined
  }

  private async dispatchInvokeAsync (context: TurnContext): Promise<InvokeResponse | undefined> {
    const name = context.activity.name
    const handler = name ? this.invokeHandlers.get(name.toLowerCase()) : undefined
    if (!handler) {
      return { status: 501 }
    }

    return withActiveSpan('botas.handler', async (handlerSpan) => {
      handlerSpan?.setAttribute('handler.type', name ?? context.activity.type)
      handlerSpan?.setAttribute('handler.dispatch', 'invoke')
      try {
        return await handler(context)
      } catch (err) {
        getMetrics()?.handlerErrors.add(1, { 'activity.type': 'invoke' })
        throw new BotHandlerException(
          `Invoke handler for "${name}" threw an error`,
          err,
          context.activity
        )
      }
    })
  }

  private async runPipelineAsync (activity: CoreActivity): Promise<InvokeResponse | undefined> {
    const context = createTurnContext(this, activity)
    let invokeResponse: InvokeResponse | undefined
    let index = 0
    const metrics = getMetrics()
    const next = async (): Promise<void> => {
      if (index < this.middlewares.length) {
        const mw = this.middlewares[index++]!
        let nextPromise: Promise<void> | undefined
        const trackedNext = (): Promise<void> => {
          nextPromise = next()
          return nextPromise
        }
        const mwStart = Date.now()
        const mwName = mw.name || `middleware-${index - 1}`
        await withActiveSpan('botas.middleware', async (mwSpan) => {
          mwSpan?.setAttribute('middleware.name', mwName)
          mwSpan?.setAttribute('middleware.index', index - 1)
          try {
            await mw(context, trackedNext)
          } catch (err) {
            // Suppress the detached next() rejection — the middleware error takes priority
            if (nextPromise) await nextPromise.catch(() => {})
            throw err
          } finally {
            metrics?.middlewareDuration.record(Date.now() - mwStart, { 'middleware.name': mwName })
          }
        })
        if (nextPromise) await nextPromise
      } else {
        invokeResponse = await this.handleCoreActivityAsync(context)
      }
    }
    try {
      await next()
    } catch (err) {
      if (err instanceof BotHandlerException) throw err
      // #67: Propagate middleware errors with original message
      throw err
    }
    return invokeResponse
  }
}

/**
 * Parse a JSON string while blocking prototype-pollution keys
 * (`__proto__`, `constructor`, `prototype`).
 */
function safeJsonParse (body: string): unknown {
  return JSON.parse(body, (key, value) => {
    if (key === '__proto__' || key === 'constructor' || key === 'prototype') {
      return undefined
    }
    return value
  })
}

/** Maximum allowed text length for an activity. */
const MAX_ACTIVITY_TEXT_LENGTH = 50_000

/** Maximum number of entities allowed on a single activity. */
const MAX_ENTITIES_COUNT = 100

/** Maximum number of attachments allowed on a single activity. */
const MAX_ATTACHMENTS_COUNT = 50

/**
 * Thrown when an incoming activity fails structural validation (missing or
 * empty required fields, oversized payloads, etc.).
 *
 * The HTTP layer should return `400 Bad Request` when this error is caught.
 */
export class ActivityValidationError extends Error {
  constructor (message: string) {
    super(message)
    this.name = 'ActivityValidationError'
  }
}

function assertCoreActivity (value: unknown): asserts value is CoreActivity {
  if (typeof value !== 'object' || value === null) {
    throw new ActivityValidationError('CoreActivity must be a JSON object')
  }
  const a = value as Record<string, unknown>
  if (typeof a['type'] !== 'string' || !a['type']) {
    throw new ActivityValidationError('Missing required field: type')
  }
  if (typeof a['serviceUrl'] !== 'string' || !a['serviceUrl']) {
    throw new ActivityValidationError('Missing required field: serviceUrl')
  }
  if (
    typeof a['conversation'] !== 'object' ||
    a['conversation'] === null ||
    typeof (a['conversation'] as Record<string, unknown>)['id'] !== 'string' ||
    !(a['conversation'] as Record<string, unknown>)['id']
  ) {
    throw new ActivityValidationError('Missing required field: conversation.id')
  }
  // #76: Validate activity field sizes to prevent abuse
  if (typeof a['text'] === 'string' && a['text'].length > MAX_ACTIVITY_TEXT_LENGTH) {
    throw new ActivityValidationError(`CoreActivity text exceeds maximum length of ${MAX_ACTIVITY_TEXT_LENGTH}`)
  }
  if (Array.isArray(a['entities']) && a['entities'].length > MAX_ENTITIES_COUNT) {
    throw new ActivityValidationError(`CoreActivity entities array exceeds maximum size of ${MAX_ENTITIES_COUNT}`)
  }
  if (Array.isArray(a['attachments']) && a['attachments'].length > MAX_ATTACHMENTS_COUNT) {
    throw new ActivityValidationError(`CoreActivity attachments array exceeds maximum size of ${MAX_ATTACHMENTS_COUNT}`)
  }
}

/** Maximum allowed request body size in bytes (1 MB). */
const MAX_BODY_BYTES = 1 * 1024 * 1024

/**
 * Thrown when an incoming request body exceeds the maximum allowed size (1 MB).
 *
 * The HTTP layer returns a `413` status code when this error is caught.
 */
export class RequestBodyTooLargeError extends Error {
  constructor () {
    super('Request body too large')
    this.name = 'RequestBodyTooLargeError'
  }
}

function readBody (req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = []
    let total = 0
    req.on('data', (chunk: Buffer) => {
      total += chunk.length
      if (total > MAX_BODY_BYTES) {
        req.destroy()
        return reject(new RequestBodyTooLargeError())
      }
      chunks.push(chunk)
    })
    req.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
    req.on('error', reject)
  })
}

function resolveBotApplicationOptions (options: BotApplicationOptions): BotApplicationOptions {
  return {
    clientId: options.clientId ?? process.env['CLIENT_ID'],
    clientSecret: options.clientSecret ?? process.env['CLIENT_SECRET'],
    tenantId: options.tenantId ?? process.env['TENANT_ID'],
    token: options.token,
    managedIdentityClientId:
      options.managedIdentityClientId ??
      (process.env['MANAGED_IDENTITY_CLIENT_ID'] as BotApplicationOptions['managedIdentityClientId']),
  }
}
